File: algorithm/Aha/chpater5/bfs.pl
#!/usr/bin/env perl 
#===============================================================================
#         FILE: bfs.pl
#       AUTHOR: Phoenix Ikki (liuxueyang.github.io), liuxueyang457@gmail.com
# ORGANIZATION: Hunan University
#      CREATED: 2016/03/12 09时42分37秒
#===============================================================================

use strict;
use warnings;
use utf8;
use 5.014;

my ($rec_row, $rec_column) = (1000+1, 1000+1);
my $point_count = 0;
my ($edges_count, $i, $j);
my (@rec, @bfs_queue, @already_visit);

sub bfs {
    my ($head, $tail, $i, $start);
    $head = $tail = 1;
    $start = shift;
    $bfs_queue[$head] = $start;
    $already_visit[$start] = 1;
    ++$tail;

    while ($head < $tail) {
        print $bfs_queue[$head], ' ';
        for (1..$point_count) {
            if ($rec[$_][$bfs_queue[$head]] == 1 &&
                $already_visit[$_] == 0) {
                $bfs_queue[$tail] = $_;
                $already_visit[$_] = 1;
                ++$tail;
            }
        }
        ++$head;
    }
}


($point_count, $edges_count) = split / /, <>;

for $i (1..$rec_row) {
    $rec[$i][$_] = 0 for (1..$rec_column);
}

$bfs_queue[$_] = 0 for (1..(1000_000+1));
$already_visit[$_] = 0 for (1..(1000_000+1));

for (1..$edges_count) {
    my ($row, $column) = split / /, <>;
    $rec[$row][$column] = 1;
    $rec[$column][$row] = 1;
}

bfs(1);

File: algorithm/Aha/chpater5/bfs.c
/*
 * =====================================================================================
 *       Filename : bfs_1.c
 *    Description : 
 *    Version     : 
 *        Created : 03/12/16 07:08
 *         Author : Liu Xue Yang (LXY), liuxueyang457@163.com
 *         Motto  : Suicide is Painless
 * =====================================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const int ROW = 1000 + 1;
const int COLUMN = 1000 + 1;
int rec[ROW][COLUMN];
int point_count;
int bfs_queue[1000 * 1000 + 1];
int already_visit[1000 * 1000 + 1];

void bfs(int start) {
  int head = 1, tail = 1;

  bfs_queue[head] = start;
  already_visit[start] = 1;
  ++tail;

  while(head < tail) {
    printf("%d ", bfs_queue[head]);

    for(int i = 1; i <= point_count; ++i) {
      if (rec[i][bfs_queue[head]] == 1 &&
          already_visit[i] == 0) {
        bfs_queue[tail] = i;
        // mark the point visited HERE! NOT before
        // the for loop :(
        already_visit[i] = 1;
        ++tail;
      }
    }
    
    ++head;
  }
}

int main(void)
{
  freopen("bfs.txt", "r", stdin);
  
  int edges_count, row, column;
  
  scanf("%d %d", &point_count, &edges_count);
  memset(rec, 0, sizeof(rec));
  memset(bfs_queue, 0, sizeof(bfs_queue));
  memset(already_visit, 0, sizeof(already_visit));

  for (int i = 1; i <= edges_count; ++i) {
    scanf("%d %d", &row, &column);
    rec[row][column] = 1;
    rec[column][row] = 1;
  }

  bfs(1);
  
  return 0;
}

